server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -275,88 +216,78 @@ private void setEnabled(boolean enabled) {
-        // Watermark is expressed in terms of used data, but we need "free" data watermark
-        this.freeDiskThresholdFloodStage = 100.0 - thresholdPercentageFromWatermark(floodStageRaw);
-        this.freeBytesThresholdFloodStage = thresholdBytesFromWatermark(
-            floodStageRaw,
-            CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey()
-        );
+    private void setFloodStageWatermark(RelativeByteSizeValue floodStage) {
+        this.floodStageWatermark = floodStage;
     }
 
-    private void setFrozenFloodStage(RelativeByteSizeValue floodStage) {
-        this.frozenFloodStage = floodStage;
+    private void setFrozenFloodStageWatermark(RelativeByteSizeValue floodStage) {
+        this.frozenFloodStageWatermark = floodStage;
     }
 
     private void setFrozenFloodStageMaxHeadroom(ByteSizeValue maxHeadroom) {
         this.frozenFloodStageMaxHeadroom = maxHeadroom;
     }
 
-    /**
-     * Gets the raw (uninterpreted) low watermark value as found in the settings.
-     */
-    public String getLowWatermarkRaw() {
-        return lowWatermarkRaw;
+    private ByteSizeValue getFreeBytesThreshold(ByteSizeValue total, RelativeByteSizeValue watermark, ByteSizeValue maxHeadroom) {
+        // If bytes are given, they can be readily returned as free bytes. If percentages are given, we need to calculate the free bytes.
+        if (watermark.isAbsolute()) {
+            return watermark.getAbsolute();
+        }
+        return ByteSizeValue.ofBytes(total.getBytes() - watermark.calculateValue(total, maxHeadroom).getBytes());
     }
 
-    /**
-     * Gets the raw (uninterpreted) high watermark value as found in the settings.
-     */
-    public String getHighWatermarkRaw() {
-        return highWatermarkRaw;
+    public ByteSizeValue getFreeBytesThresholdLowStage(ByteSizeValue total) {
+        return getFreeBytesThreshold(total, lowStageWatermark, ByteSizeValue.ofBytes(-1));
     }
 
-    public Double getFreeDiskThresholdLow() {
-        return freeDiskThresholdLow;
+    public ByteSizeValue getFreeBytesThresholdHighStage(ByteSizeValue total) {
+        return getFreeBytesThreshold(total, highStageWatermark, ByteSizeValue.ofBytes(-1));
     }
 
-    public Double getFreeDiskThresholdHigh() {
-        return freeDiskThresholdHigh;
+    public ByteSizeValue getFreeBytesThresholdFloodStage(ByteSizeValue total) {
+        return getFreeBytesThreshold(total, floodStageWatermark, ByteSizeValue.ofBytes(-1));
     }
 
-    public ByteSizeValue getFreeBytesThresholdLow() {
-        return freeBytesThresholdLow;
+    public ByteSizeValue getFreeBytesThresholdFrozenFloodStage(ByteSizeValue total) {
+        return getFreeBytesThreshold(total, frozenFloodStageWatermark, frozenFloodStageMaxHeadroom);
     }
 
-    public ByteSizeValue getFreeBytesThresholdHigh() {
-        return freeBytesThresholdHigh;
-    }
+    private ByteSizeValue getMinimumTotalSizeForBelowWatermark(
```

Comment:
This is only used for the low watermark (at least for now). Do you need it to be this general or can we inline it and then just add the headroom parameter in the next PR?